File: 967-minimum-falling-path-sum/minimum-falling-path-sum.java
class Solution {
    int[][] dp;
    public int minFallingPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;

        int[] prev = new int[n];

        for(int j = 0 ; j < n ; j++){
            prev[j] = grid[m-1][j];
        }

        for(int i = m - 2 ; i >= 0 ; i--){
            int[] curr = new int[n];
            for(int j = n - 1 ; j >= 0 ; j--){
                int down = grid[i][j] + prev[j];
                int leftDiag = Integer.MAX_VALUE;
                int rightDiag = Integer.MAX_VALUE;

                if(j > 0){
                    leftDiag = grid[i][j] + prev[j-1];
                } 

                if(j < n-1){
                    rightDiag = grid[i][j] + prev[j+1];
                }

                curr[j] = Math.min(down , Math.min(leftDiag , rightDiag));
            }

            prev = curr;

        }

        int mini = Integer.MAX_VALUE;
        for(int j = 0 ; j < n ; j++){
            mini = Math.min(mini , prev[j]);
        }
        return mini;
    }

   
}

File: 967-minimum-falling-path-sum/minimum-falling-path-sum.py
class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        n = len(matrix)
        res = float('inf')

        dp = matrix[n-1][:]
        curr = [0] * n
        
        for i in range(n-2 , -1 , -1):
            for j in range(n-1, -1 , -1):
                curr[j] = matrix[i][j] + min(
                    dp[j],
                    dp[j-1] if j-1 >= 0 else float('inf'),
                    dp[j+1] if j+1 < n else float('inf') 
                )
            
            dp = curr[:]


        for j in range(n):
            res = min(res , dp[j])
        
        return res
        